### ISSUE TITLE
Make GPUDevice.destroy() a queued operation

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#2599]
It came up in today's editors' call that GPUDevice.destroy() should probably be a queued operation, i.e. occurring after previous operations on the device timeline. This would allow patterns like:
- Render to a canvas and destroy before waiting for it to be presented
- Do some work, issue a mapAsync for readback, and destroy before waiting for the map to complete

I am pretty sure this can be a post-V1 change, but we should briefly discuss before deciding that.

**Comment by kainino0x** [comment]
For the purposes of testing, this also means device destroy() would never interrupt currently queued work the way a real device lost could. We may want to add a way of testing immediate device loss, but browsers could do this in their own tests with testing-only APIs, or by terminating the GPU process or WebGPU device by internal means.

**Comment by Kangz** [comment]
> It came up in today's editors' call that GPUDevice.destroy() should probably be a queued operation, i.e. occurring after previous operations on the device timeline. This would allow patterns like:

Did you mean a device timeline operation, but that implicitly waits for all submitted work to be done?

**Comment by kainino0x** [comment]
@Kangz yeah, should have been more precise.

**Comment by greggman** [comment]
So let's say I have a memory hungry (lots of large buffers and textures) WebGPU app and for some reason I want to destroy the GPUDevice and create a new one. What are the steps I need to take to do this safely. In other words. Imagine there is 1meg of memory available and I do this

```
const device1= await adapter.requestDevice();
const buffer1 = device1.createBuffer({size: 1024 * 1024, ... });
device1.destroy();

const device2 = await adapter.requestDevice();
const buffer2 = device2.createBuffer({size: 1024 * 1024, ... });
device2.destroy();
```

It sounds like if device.destroy is queued then there's very little guarantee the second buffer creation will succeed vs if it's synchronous.

To put it another way, what's the correct way to use the API if device.destroy is queued?

**Comment by kainino0x** [comment]
The proposal is essentially to make device.destroy work the same way as buffer.destroy, where it is well ordered with regard to other operations (e.g. `submit(); b1.destroy()` is OK). So I'd like your example to be analogous to a scenario where you have just one device and do `b1.destroy(); b2 = createBuffer()`.

Currently the spec describes "device timelines" as being per-device, but the spec model could change so there's just one global device timeline and this is well defined. (Implementations would be able to parallelize things as long as it behaves "as if" ordered according to the spec.)

**Comment by greggman** [comment]
>  but the spec model could change so there's just one global device timeline and this is well defined. (Implementations would be able to parallelize things as long as it behaves "as if" ordered according to the spec.)

That would probably work. My concern is probably better illustrated with

```
const device1 = await adapter.requestDevice();
const buffer1 = device1.createBuffer({size: 1024 * 1024, ... });
context.configure(...);
const commandEncoder = device.createCommandEncoder();
const passEncoder = commandEncoder.beginRenderPass(renderPassDesc);
...
passEncoder.draw(6, 1);
passEncoder.endPass();
device1.queue.submit([commandEncoder.finish()]);
device1.destroy();

const device2 = await adapter.requestDevice();
const buffer2 = device2.createBuffer({size: 1024 * 1024, ... });
device2.destroy();
```

So now what happens.... does the submit execute? Will the second creation wait ? It seems unlikely those 2 things are connected as its normal or expected to be able to create a new device at any time but the resources are being used.

It seems like the spec would have to say something like "if a device destruction is queued, requestDevice must not return a new device until the queued destruction is complete".  

Either that, or device.destroy returns a promise or you use popErrorScope to get one that's speced not to resolve until destruction is complete.

**Comment by kainino0x** [comment]
<details><summary>WebGPU meeting 2022-06-22</summary>

- KN: thinking about post-V1. Only makes invalid programs valid. Proposal was: destroying the device doesn't jump ahead of already-issued work. Submits, maps, render-to-canvas will complete.
- KN: don't think this is critical. But would clean up a confusing area about ordering.
- KG: how observable is this?
- KN: observable if you e.g. render to canvas, then destroy the device. Different impls might have different behaviors. A weird ordering issue.
- KG: if you destroy device, is visual output guaranteed?
- KN: not really spec'd now. De facto now, no. Spec doesn't say enough to guarantee it. Impl could go either way.
- RC: Is destroying normal textures a queued operation? If so, this should be queued too.
- KN: destroy w.r.t. spec is all about validation. Also true of device.destroy(). If you go through spec & steps - this is probably already true. Not really the intent though. Will take editorial work to figure out how it fits into spec. At high level - if we think this is a good idea, we should decide this and take care of it in editing.
- KN: our impl for example - end of frame, we transfer a resource from WebGPU to compositor for display. If you did submit, then destroy the device, then end-of-frame happens after that - is the thing dead, and you can't get resource over to compositor? Or allocated outside device, so it's still alive? This ambiguity is why it could be different in different browsers.
- KG: to me - dependent on destroying a device. Does anything show up on screen? If we don't say anything about that, this doesn't matter. Even if we say this waits - still unspecified after you finish waiting.
- KN: for that situation I think we should have tighter specification here. No mention of that now.
- KG: different question - if you destroy device, then call canvas.toDataUrl() - do you get anything? Do we say, you don't have a device any more so you get nothing? Maybe if you destroy the device, we take one final snapshot of the canvas, even if just in an ImageData. That'd give us some resolution here. Either you asked for device to be lost, and you still have canvas output - or you have device lost, and the canvas was unrecoverable. Could "try" in the cooperative case. "Implementations should preserve the final contents of the canvas rendering context"?
- KN: makes sense. Also question of: if you render frame, wait for end of frame, then you destroy device - does canvas disappear, or leave the image behind? Spec currently should leave the image there, think that's the correct behavior. Given that, think we should extend that behavior here. Destroy before present has similar behavior.
- KN: device loss trickier, can happen any time. Need to think about presentation interactions.
- KG: think I am more convinced that GpuDevice.destroy() waits for stuff to happen before it finishes.
- KN: spec-wise with 3 timelines - destroy() goes to device timeline to execute. Do we wait for queue timeline before destroying device, or cut things off? Ordering violation between timelines if so. But we don't have good explanation of getting data out of queue timeline.
- KG: is this device.destroy(), or device.release()? Don't just knock things down for no reason. Think we really want release(), not destroy().
- KN: if you destroy a texture while something in queue timeline, things don't blow up.
- JB: making destroy() the default mean that some impls will do it now, and some impls will wait. Useful to have cancel() operation? Shouldn't be the default. Default should be to wait for things to complete. No race conditions.
- KN: makes sense.
- KG: say we're moving that way?
- KN: sgtm. We'll figure out details in editing.

</details>
